File: python/020. Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        paraMap = {')': '(', ']': '[', '}': '{'}
        q = [None]
        for c in s:
            if c in paraMap and q[-1] == paraMap[c]:
                q.pop()
            else:
                q.append(c)
        return len(q) == 1

File: C++/020. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) 
    {
        int n = s.length();
        if (n == 0) return true;
        stack<char> myStack;
        for (int i = 0; i < n; i++)
        {
            char cur = s[i];
            if (cur == '(' || cur == '[' || cur == '{')
            {
                myStack.push(cur);
            }
            else if (!myStack.empty() )
            {
                char old  = myStack.top();
                if ((old == '(' && cur == ')') || (old == '[' && cur == ']') || (old == '{' && cur == '}') )
                {
                    myStack.pop();
                }
                else
                {
                    return false;
                }
            }
            else
            {
                return false;
            }
        }
        return myStack.empty();
    }
};
